### PR TITLE
Feature/1149 pathfinder v2

### PR DESCRIPTION
**Pull request by mitzimorris**
#### Submisison Checklist

- [x] Run tests: `./runCmdStanTests.py src/test`
- [x] Declare copyright holder and open-source license: see below

#### Summary:

adds logic to CmdStan's `command` method and argument parser to handle calls to Pathfinder.

#### Intended Effect:

expose Pathfinder.

#### How to Verify:

unit tests

#### Documentation:

will be done in a separate PR

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company):   Columbia University

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

### CODE REVIEW

**Comment by mitzimorris** [comment]
Initial commit; adds logic for Pathfinder.
For optional diagnostic files, constructs `unique_stream_writer` and/or `json_writer` objects with nullptr to output stream.

Current branch `stan/multi-path` doesn't yet play nicely with json_writer.

**Comment by mitzimorris** [comment]
this is ready for review.

File: src/test/interface/pathfinder_test.cpp
Comment (WardBrian) [inline src/test/interface/pathfinder_test.cpp:36]:
```suggestion
     << " method=pathfinder num_psis_draws=40";
      run_command_output out = run_command(ss.str());
```

File: src/cmdstan/arguments/arg_variational_iter.hpp
Comment (WardBrian) [inline src/cmdstan/arguments/arg_variational_iter.hpp:1]: All these small lexical cast changes seem fine but may be better in a separate PR

File: src/cmdstan/command.hpp
Comment (WardBrian) [inline src/cmdstan/command.hpp:149]: It appears that the `validate_multi_chain_config` function is no longer called anywhere. Either its functionality was replaced (in which case we can delete it from the helpers file) and I can't find it, or we lost this check somewhere and need to put it back

File: src/cmdstan/command.hpp
Comment (WardBrian) [inline src/cmdstan/command.hpp:261]: Is there a reason why the `save_iterations` argument isn't used (in place of the `false` on this line)?

File: src/cmdstan/command.hpp
Comment (WardBrian) [inline src/cmdstan/command.hpp:286]: Similarly, `save_iterations` instead of hard-coded `true`?

File: src/cmdstan/command.hpp
Comment (WardBrian) [inline src/cmdstan/command.hpp:288]: Is this passing the same JSON diagnostic writer as both the writer for the first pathfinder and for the overall multi-pathfinder?

File: src/test/interface/pathfinder_test.cpp
Comment (WardBrian) [inline src/test/interface/pathfinder_test.cpp:97]: Can you also add a test for multi-path for both kinds of diagnostic file?

File: src/test/interface/pathfinder_test.cpp
Comment (WardBrian) [inline src/test/interface/pathfinder_test.cpp:1]: There are several sub-arguments to pathfinder that are not tested by this file (e.g. `save_iterations`, `num_draws`, etc) that would be good to have some basic tests for

**Review by WardBrian — CHANGES_REQUESTED** [review]

File: src/cmdstan/command.hpp
Comment (mitzimorris) [reply src/cmdstan/command.hpp:149]: good catch.  we should have this - moved too much stuff around.  added back.

File: src/cmdstan/command.hpp
Comment (mitzimorris) [reply src/cmdstan/command.hpp:261]: no.  fixed.

File: src/cmdstan/command.hpp
Comment (WardBrian) [reply src/cmdstan/command.hpp:149]: I noticed another function in `helper` called `allow_multichain` which seems to do mostly the same thing (+ some pathfinder logic), but doesn't look exactly the same

File: src/cmdstan/command.hpp
Comment (mitzimorris) [reply src/cmdstan/command.hpp:288]: no.   diagnostic writers are only on the individual single-path pathfinders.  the overall multi-pathfinder writer is the importance resampled draws from the single-path pathfinders.  no gradients involved.

File: src/test/interface/pathfinder_test.cpp
Comment (mitzimorris) [reply src/test/interface/pathfinder_test.cpp:97]: there's only one kind of diagnostic file.

File: src/cmdstan/command.hpp
Comment (mitzimorris) [reply src/cmdstan/command.hpp:149]: that's used by another helper function - `get_num_chains`.  added back `validate_multi_chain_config` call, also using `get_num_chains` since with pathfinder,  `num_paths` also determines `num_chains`.

File: src/cmdstan/command.hpp
Comment (mitzimorris) [reply src/cmdstan/command.hpp]: added dummy_writer.

File: src/test/interface/pathfinder_test.cpp
Comment (mitzimorris) [reply src/test/interface/pathfinder_test.cpp:1]: added tests for max_lbfgs_iters and num_draws.

File: src/test/interface/pathfinder_test.cpp
Comment (WardBrian) [inline src/test/interface/pathfinder_test.cpp:139]: This appears to be failing

File: src/cmdstan/command_helper.hpp
Comment (WardBrian) [inline src/cmdstan/command_helper.hpp:698]: I don't believe "unit" is a valid metric - "unit_e"?

File: src/cmdstan/command_helper.hpp
Comment (WardBrian) [inline src/cmdstan/command_helper.hpp:696]: The second half of this function is repeating the logic of the `validate_multi_chain_config` function but just returns a boolean rather than raising an error. We should only be doing this one place

File: src/cmdstan/command_helper.hpp
Comment (WardBrian) [inline src/cmdstan/command_helper.hpp:726]: This is also duplicate code when `validate_multi_chain_config` is considered.

I think most likely we should just remove `allow_multichain` and this validation and use the pre-existing `validate_multi_chain_config` if we're sampling

File: src/cmdstan/command.hpp
Comment (WardBrian) [inline src/cmdstan/command.hpp:255]: This flag is still not used anywhere in the following code, which has hardcoded true/false values for the same-named service function argument

Comment (WardBrian) [reply src/cmdstan/command.hpp:255]: This needs to be tested

File: src/cmdstan/command.hpp
Comment (WardBrian) [inline src/cmdstan/command.hpp:191]: Because it's currently unusued it does not matter, but I suspect if it ever becomes used the Multi-path pathfinder service will expect a structured writer here rather than a `writer`

**Review by WardBrian — CHANGES_REQUESTED** [review]

File: src/cmdstan/command_helper.hpp
Comment (mitzimorris) [reply src/cmdstan/command_helper.hpp:698]: good catch - fixed!

File: src/cmdstan/command_helper.hpp
Comment (mitzimorris) [reply src/cmdstan/command_helper.hpp:726]: much cleaner - thanks.

File: src/cmdstan/command.hpp
Comment (mitzimorris) [reply src/cmdstan/command.hpp:255]: right.  testing this, and there's  definitely a problem here.  seems to be always saving iters, no matter what the setting is.

Comment (WardBrian) [reply src/cmdstan/command.hpp:255]: As currently written it should always save iterations for multi-path pathfinder and never save them for single path. This is assuming it is all correct in the services layer, of course

Comment (mitzimorris) [reply src/cmdstan/command.hpp:255]: this is the problem:   "save_iterations" means dump lbfgs information to diagnostic_writer.  setting "save_iterations=1" (true), only makes sense if the output sub-arg "diagnostic_file" is specified.  we shouldn't expose "save_iterations" at all; instead, set the bool value "save_iterations" to whether or not arg "diagnostic_file" is specified.

Comment (mitzimorris) [reply src/cmdstan/command.hpp:255]: removed "save_iterations" as user-facing argument.  feature tested via tests for diagnostic_file.

File: src/cmdstan/command.hpp
Comment (mitzimorris) [reply src/cmdstan/command.hpp:191]: changed to json_writer.

File: src/test/interface/pathfinder_test.cpp
Comment (mitzimorris) [reply src/test/interface/pathfinder_test.cpp]: tweaked test logic - should reliably pass.

**Comment by mitzimorris** [comment]
changes made - one more time?

**Review by WardBrian — COMMENTED** [review]
A few minor questions/style comments but otherwise I think this is looking pretty good.

While playing around I noticed that the CSV for multi-pathfinder appears to be incorrect (the header has the columns in the wrong order). So that needs to be fixed before we merge this (ping @SteveBronder)

File: src/cmdstan/command.hpp
Comment (mitzimorris) [reply src/cmdstan/command.hpp:277]: evidently - posteriordb has some csv files that have suffix ".tmp".

Comment (WardBrian) [reply src/cmdstan/command.hpp:277]: Today I learned. Sounds reasonable.

Was there a discussion anywhere of the `_pathfinder` naming convention?

File: src/cmdstan/command_helper.hpp
Comment (mitzimorris) [reply src/cmdstan/command_helper.hpp:813]: when running pathfinder, create vector of csv writers with nullptrs and when running sampler create vector of json writers with nullptrs?

Comment (WardBrian) [reply src/cmdstan/command_helper.hpp:813]: I think that is the simplest thing, and I believe it will only take 2*num_chains bytes worth of extra space compared to what is being done now

File: src/cmdstan/command.hpp
Comment (mitzimorris) [reply src/cmdstan/command.hpp:277]: @SteveBronder

> Was there a discussion anywhere of the _pathfinder naming convention?

**Comment by WardBrian** [comment]
Another question. Right now, if one runs ADVI, you get the following as part of the output:

```

------------------------------------------------------------
EXPERIMENTAL ALGORITHM:
  This procedure has not been thoroughly tested and may be unstable
  or buggy. The interface is subject to change.
------------------------------------------------------------
```

Is there a desire or need to have a similar warning here?

**Comment by mitzimorris** [comment]
I would hope not.  have we tested this pathfinder algorithm on PosteriorDB and does it get the same (or similar) results as the R implementation evaluations reported in the Pathfinder paper?   @SteveBronder ?

**Comment by WardBrian** [comment]
I was primarily asking because of @bob-carpenter's remark in the original issue (https://github.com/stan-dev/stan/issues/3119) 
>This is all going in the experimental folder and will be marked as such like ADVI when run.

**Comment by mitzimorris** [comment]
yes, I noticed the remark in the initial issue.  personally, I would prefer one less directory in the path, but that's just to save keystrokes.

**Comment by mitzimorris** [comment]
this failed because downstream checks had warnings from boost::lexical_cast using "sprintf".   the fix for this is https://github.com/stan-dev/math/pull/2919 is someone wants to review and approve.

**Comment by WardBrian** [comment]
I wonder why other PRs don’t seem to fail because of that warning. I don’t think it’s a blocker for merging this at any rate

File: src/cmdstan/command_helper.hpp
Comment (WardBrian) [inline src/cmdstan/command_helper.hpp:805]: Basic C++ question: is there a reason to prefer `.resize(0)` to `.clear()`?

Comment (mitzimorris) [reply src/cmdstan/command_helper.hpp:805]: maybe?  discussion inconclusive:  https://stackoverflow.com/questions/2738967/vector-clear-vs-resize

Comment (WardBrian) [reply src/cmdstan/command_helper.hpp:805]: It sounds like in C++11 and later clear is not allowed to change the capacity, so either seems fine.

**Comment by WardBrian** [comment]
I think the only remaining question I have before approval is whether or not we believe we should be calling `stan::services::util::experimental_message` to produce a warning like ADVI

Thoughts @SteveBronder @bob-carpenter?

File: src/cmdstan/command_helper.hpp
Comment (WardBrian) [reply src/cmdstan/command_helper.hpp:805]: I believe the current test failures are actually caused by the `.remove(0)`, since `remove` requires default-constructable types. Changing it to clear() resolves the issue for me locally.

**Comment by WardBrian** [comment]
@mitzimorris you can fix the CI failure by changing `resize(0)` to `clear()` on those std::vectors. The reason linking is failing is that the unique_stream_writer has a forward-declared default constructor, but it never gets defined (this is probably a separate issue @SteveBronder should fix, but `clear()` will work and is more semantically meaningful)

File: src/cmdstan/command.hpp
Comment (SteveBronder) [inline src/cmdstan/command.hpp:270]: Instead of functions that have input/output arguments can we have this return a `std::pair<std::string, std::string>`?

File: src/cmdstan/command.hpp
Comment (SteveBronder) [inline src/cmdstan/command.hpp:276]:
```suggestion
      if (sig_figs > -1) {
        ofs->precision(sig_figs);
     }
```

File: src/cmdstan/command_helper.hpp
Comment (SteveBronder) [inline src/cmdstan/command_helper.hpp:691]: I prefer curlies around the bodies of its. It mostly helps readability and to make sure someone in the future doesn't think they are in an if when it only effects the next line

File: src/cmdstan/command_helper.hpp
Comment (SteveBronder) [inline src/cmdstan/command_helper.hpp:722]: Can you just use an std string here?

File: src/cmdstan/command_helper.hpp
Comment (SteveBronder) [inline src/cmdstan/command_helper.hpp:708]: Instead of defining `input_file` here can we just define it where it's constructed in the if statement bodies?

File: src/cmdstan/command_helper.hpp
Comment (SteveBronder) [inline src/cmdstan/command_helper.hpp:781]: Can we rewrite `make_filenames` to return an std vector of strings?

File: src/cmdstan/command_helper.hpp
Comment (SteveBronder) [inline src/cmdstan/command_helper.hpp:772]: Are all 3 of these vectors input/output arguments? If they come in as uninitialized arguments then I'd rather return a tuple or struct holding the different writers rather than have them be input / output arguments

**Review by SteveBronder — CHANGES_REQUESTED** [review]
In general looks good, though I'd like to avoid using input/output arguments so much and just pairs/tuples/structs with the objects constructed in the function

File: src/cmdstan/command_helper.hpp
Comment (mitzimorris) [reply src/cmdstan/command_helper.hpp:691]: in this case, I disagree because and if which wraps a return stmt w/out curly braces is pretty unambiguous.

File: src/cmdstan/command_helper.hpp
Comment (mitzimorris) [reply src/cmdstan/command_helper.hpp:722]: I think the filename is useful information - is that what you're asking?  or are you suggesting avoid setting up the std::stringstream and just constructing the message as argument to the `std::invalid_argument` constructor?

I would prefer to leave it as is because
+  this makes it easy to see the message
+ it's consistent with error messaging elsewhere.

File: src/cmdstan/command_helper.hpp
Comment (mitzimorris) [reply src/cmdstan/command_helper.hpp:772]: yes, they come in as uninitialized args, but creating a tuple struct doesn't make much sense here, given the calling structure for the services.  the main goal of `init_callbacks` is to move a chunk of tedious code out of the main part of the `command` function, in order to make it a little easier to see the overall logic flow.
not changing.

File: src/cmdstan/command_helper.hpp
Comment (mitzimorris) [reply src/cmdstan/command_helper.hpp]: fixed!

**Comment by mitzimorris** [comment]
requested changes made.  ready for re-re-re-review.

**Review by WardBrian — APPROVED** [review]
I’m happy with the changes, I’ll leave final review for @SteveBronder

File: src/cmdstan/command_helper.hpp
Comment (SteveBronder) [inline src/cmdstan/command_helper.hpp:774]: (optional) I think it could be nice to have a little struct for these that could be like

```c++
struct Callbacks {
  std::vector<stan::callbacks::unique_stream_writer<std::ofstream>> sample_csv;
  std::vector<stan::callbacks::unique_stream_writer<std::ofstream>> diagnostic_csv;
  std::vector<stan::callbacks::json_writer<std::ofstream>> diagnostic_json;
  explicit Callbacks(argument_parser& parser) {
    auto user_method = parser.arg("method");
    unsigned int num_chains = get_num_chains(parser);
    unsigned int id = get_arg_val<int_argument>(parser, "id");
    int sig_figs = get_arg_val<int_argument>(parser, "output", "sig_figs");

    sample_writers.reserve(num_chains);
    std::vector<std::string> output_filenames
        = make_filenames(get_arg_val<string_argument>(parser, "output", "file"),
                        ".csv", num_chains, id);
    for (int i = 0; i < num_chains; ++i) {
      auto ofs = std::make_unique<std::ofstream>(output_filenames[i]);
      if (sig_figs > -1)
        ofs->precision(sig_figs);
      sample_writers.emplace_back(std::move(ofs), "# ");
    }

    diag_json_writers.reserve(num_chains);
    diag_csv_writers.reserve(num_chains);
    // create no-op writers by default
    for (int i = 0; i < num_chains; ++i) {
      diag_json_writers.emplace_back(
          stan::callbacks::json_writer<std::ofstream>());
    }
    for (int i = 0; i < num_chains; ++i) {
      diag_csv_writers.emplace_back(nullptr, "# ");
    }
    // create json, csv writers as needed.
    std::string diagnostic_file
        = get_arg_val<string_argument>(parser, "output", "diagnostic_file");
    if (!diagnostic_file.empty()) {
      std::vector<std::string> diag_filenames;
      if (user_method->arg("pathfinder")) {
        diag_json_writers.clear();
        diag_filenames = make_filenames(diagnostic_file, ".json", num_chains, id);
        for (int i = 0; i < num_chains; ++i) {
          auto ofs = std::make_unique<std::ofstream>(diag_filenames[i]);
          if (sig_figs > -1)
            ofs->precision(sig_figs);
          stan::callbacks::json_writer<std::ofstream> jwriter(std::move(ofs));
          diag_json_writers.emplace_back(std::move(jwriter));
        }
      } else {
        diag_csv_writers.clear();
        diag_filenames = make_filenames(diagnostic_file, ".csv", num_chains, id);
        for (int i = 0; i < num_chains; ++i) {
          auto ofs = std::make_unique<std::ofstream>(diag_filenames[i]);
          if (sig_figs > -1)
            ofs->precision(sig_figs);
          diag_csv_writers.emplace_back(std::move(ofs), "# ");
        }
      }
    }
  }
};
````

Because then we call set this up like 

```c++
Callbacks writers(parser);
```

**Review by SteveBronder — APPROVED** [review]
One optional comment but lgmt!!

**Comment by mitzimorris** [comment]
current implementation of Pathfinder has default num_draws set to 2000.  why? @bob-carpenter - opinion?
NUTS-HMC runs 2000 iterations by default becuase there are 1000 warmup iters and 1000 sampling iters.

**Comment by bob-carpenter** [comment]
Our default for HMC produces 1000 warmup iterations and 1000 draws.  That's why I suggested 1000 draws as the default number of Pathfinder draws.   1024 draws from the posterior would bring MCMC standard error down to about std-deviation / 32.  But we're not even getting true posterior draws here and the importance ratios is a much better indicator of error than sample size.

**Comment by SteveBronder** [comment]
I think we just need to fix the merge conflict and we are good to go!

**Comment by mitzimorris** [comment]
@syclik @WardBrian, @serban-nicusor-toptal - the issue here is boost::lexical_cast.   proposed solution - update version of boost library?

**Comment by WardBrian** [comment]
My proposed solution for this PR is just to merge - we know about the warning, and everything else passed as-is

Re, updating boost: https://github.com/stan-dev/math/issues/2856

**Pull request merged**
